lgc/builder/BuilderImpl.cpp

```diff
@@ -503,36 +505,115 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+      // Old version of the code
+      desc = CreateIntrinsic(Intrinsic::amdgcn_waterfall_last_use, descTy, {waterfallBegin, desc}, nullptr, instName);
+#else
+      desc = CreateIntrinsic(useVgprForOperands ? Intrinsic::amdgcn_waterfall_last_use_vgpr
+                                                : Intrinsic::amdgcn_waterfall_last_use,
+                             descTy, {waterfallBegin, desc}, nullptr, instName);
+#endif
+    }
+    // Replace the descriptor operand in the buffer/image operation.
+    nonUniformInst->setOperand(operandIdx, desc);
   }
+  return waterfallBegin;
+}
 
-  return true;
+Value *BuilderImpl::emitWaterfallBeginForScalarizedLoops(Instruction *nonUniformInst, Instruction *firstIndexInst,
+                                                         ArrayRef<unsigned> operandIdxs,
+                                                         ArrayRef<Value *> nonUniformIndices, bool useVgprForOperands,
+                                                         const Twine &instName) {
+  assert(firstIndexInst);
+  auto descTy = firstIndexInst->getType();
+  SetInsertPoint(nonUniformInst);
+  Instruction *waterfallBegin = CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, descTy,
+                                                {ConstantInt::get(getInt32Ty(), 0), firstIndexInst}, nullptr, instName);
+  Instruction *readFirstLane = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {descTy, descTy},
+                                               {waterfallBegin, firstIndexInst}, nullptr, instName);
+  // Find the loads that should be pushed in the loop and copy them inside the waterfall loop.
+  for (unsigned operandIdx : operandIdxs) {
+    Value *desc = nonUniformInst->getOperand(operandIdx);
+    Value *OrigLoad = nullptr;
+    Value *OrigAnd = nullptr;
+    Value *OrigGep = nullptr;
+    Value *OrigExtract = nullptr;
+    if (isa<LoadInst>(desc) && match(desc, m_Load(m_Value(OrigGep)))) {
+      Instruction *newLoad = cast<LoadInst>(desc)->clone();
+      auto *newGep = dyn_cast<GetElementPtrInst>(OrigGep)->clone();
+      newLoad->insertBefore(nonUniformInst);
+      nonUniformInst->setOperand(operandIdx, newLoad);
+      newGep->insertBefore(newLoad);
+      newGep->setOperand(1, readFirstLane);
+      newLoad->setOperand(0, newGep);
+    } else if (isa<InsertElementInst>(desc) &&
+               match(desc, m_InsertElt(m_Value(OrigLoad), m_Value(OrigAnd), m_ConstantInt())) &&
+               match(OrigLoad, m_Load(m_Value(OrigGep))) && match(OrigAnd, m_And(m_Value(OrigExtract), m_Value())) &&
+               match(OrigExtract, m_ExtractElt(m_Specific(OrigLoad), m_Value()))) {
+
+      // This case is for gfx 10.3.2
+      auto *insertOp = cast<InsertElementInst>(desc);
+      auto *loadOp = dyn_cast<LoadInst>(OrigLoad);
+      auto *gepOp = dyn_cast<GetElementPtrInst>(OrigGep);
+      auto *AndOp = dyn_cast<Instruction>(OrigAnd);
+      auto *extractOp = dyn_cast<ExtractElementInst>(OrigExtract);
```

Comment:
These should all just be `cast`. No reason to use `dyn_cast` after previously doing all the match checks.

Reply:
Thank you for pointing out :) Not needed any more.